### ISSUE TITLE
[RLlib] Feature request: reward-hacking onset monitoring hooks

### ISSUE THREAD

**Issue by Aarav500** [ray-project/ray#64655]
## Feature request
Add optional monitoring hooks/callbacks to RLlib's training loop (PPO and similar) that track signals correlated with reward-hacking onset — KL-vs-reference acceleration, entropy-collapse trend, and advantage-distribution drift (Wasserstein vs. a rolling baseline). Related code: https://github.com/Aarav500/flight-recorder (Apache-2.0), which implements these extractors already.

## Motivation
Reward hacking is a common failure mode in RL training, and RLlib currently has no standard callback for monitoring it. This comes from work on two reward-hacking-detection benchmarks (RHOB, Flight Recorder) — one relevant finding: these signals reliably audit hacking after the fact but don't reliably give early warning at a usable false-positive rate (paper: "Audit or Early Warning? A Benchmark for Online Detectors of Reward Hacking").

Given Ray/RLlib's scale, wanted to raise this as a proposal first rather than a PR — happy to share methodology/code if there's interest in scoping an RLlib-style callback.

**Comment by Aarav500** [comment]
Following up briefly — since filing this, the underlying benchmark (RHOB) has grown from 14 to 18 families and a synthetic RLHF reward-model-overoptimization setting is landing this week (23 families total). If RLlib has interest in reward-hacking-onset monitoring hooks (framework-agnostic core signals already exist in a companion project, Flight Recorder — KL-drift, entropy-collapse, advantage-drift), happy to help scope an RLlib-specific callback. Just wanted to check back in once rather than let it go silent.

**Comment by ArturNiederfahrenhorst** [comment]
Hi @Aarav500 ! Thanks for opening this.
It's an exciting topic and we'd like to see a community contribution here.
The shape of that would be an example in `rllib/examples/` plus a quick but effictive test showing this technique together with PPO.

**Comment by Aarav500** [comment]
Thanks Artur — that's a clear, concrete shape and I'm glad to build to it rather than guess at scope.

To confirm I've got it right: an example under `rllib/examples/` that wraps PPO with a monitoring callback emitting Flight Recorder's core signals (KL-acceleration vs. the frozen reference policy, entropy collapse, advantage-distribution drift), plus a test exercising it end-to-end on a small PPO run so it's demonstrably not just plumbing that looks right on paper.

I'm mid-way through some other benchmark work right now, so I can't commit to a specific date, but I'll build this as a real PR against `rllib/examples/` (not just talk about it) and open it here once it's ready, rather than leave this open-ended. If there's a minimum PPO config/env you'd want the example to target (CartPole-scale vs. something meatier) let me know, otherwise I'll pick something small and fast to keep the test cheap to run in CI.

**Comment by rehan243** [comment]
fwiw, I've worked on similar reward-hacking detection stuff before, and KL-vs-reference acceleration + advantage drift are solid signals to track. One thing I found useful when trying to catch this early was normalizing the advantage distributions to account for scale changes over time, otherwise, the Wasserstein distance would sometimes spike just because of reward scaling quirks, not actual hacking behavior.

If you're thinking about adding this as a callback, you could piggyback on the existing `on_train_result` hook in RLlib. I used something like this for logging KL changes during PPO training:

```python
def reward_hacking_monitor(trainer, result):
    kl = result["info"]["learner"]["default_policy"]["kl"]
    advantage_mean = np.mean(result["hist_stats"]["policy_loss"])
    result["custom_metrics"]["kl"] = kl
    result["custom_metrics"]["advantage_mean"] = advantage_mean

trainer = PPOTrainer(config={"callbacks": {"on_train_result": reward_hacking_monitor}})
```

You'd obviously need to extend this with the entropy and Wasserstein stuff, but the framework's already there. The tricky part I ran into was setting reasonable thresholds for alarms, like you said, tuning for good false-positive rates is tough.

**Comment by Aarav500** [comment]
@rehan243 good catch, and worth being precise about since it's a real gap, not a hypothetical: I went and checked Flight Recorder's actual `AdvantageExtractor.update()` (`flightrecorder/core/advantage.py`), and it currently computes `wasserstein_distance(a, baseline)` directly on raw advantage values with zero scale normalization. So yes — a reward-scale change (or even natural scale drift over training) would show up as drift regardless of whether the underlying *shape* of the distribution actually changed, exactly the false-positive mode you described. I'll fix that (z-scoring or otherwise rescaling both `a` and `baseline` to a common scale before the Wasserstein comparison) before this becomes part of the RLlib example I'm building for @ArturNiederfahrenhorst above — didn't want to fold in a signal I know has this specific hole without flagging it.

The `on_train_result` callback shape is helpful too — that's basically the hook point I was planning to use, so thanks for saving me the trial-and-error there. Agreed threshold-tuning is the hard part; that's the actual open question this whole project is trying to answer honestly rather than paper over (our own results show these signals separate hacking from non-hacking runs well after the fact, but reliable *early* warning at a usable false-positive rate is still unresolved — see the repo's README for the actual numbers rather than me overselling it here).
